webui/react/src/shared/themes.test.ts

```diff
@@ -0,0 +1,86 @@
+import * as themes from './themes';
+import { isColor, labDistance, rgb2lab, str2rgba } from './utils/color';
+
+const supportedThemes = {
+  darkDet: themes.themeDarkDetermined,
+  darkHpe: themes.themeDarkHpe,
+  lightDet: themes.themeLightDetermined,
+  lightHpe: themes.themeLightHpe,
+};
+
+type ThemeVarPair = (keyof themes.Theme)[];
+const genPairs = (name: string): ThemeVarPair[] => {
+  const base = [
+    [ `${name}`, `${name}On` ],
+    [ `${name}`, `${name}Border` ],
+  ] as ThemeVarPair[];
+
+  // same intentsity pairs.
+  const strong = [ ...base.map(([ a, b ]) => [ `${a}Strong`, `${b}Strong` ]) ] as ThemeVarPair[];
+  const weak = [ ...base.map(([ a, b ]) => [ `${a}Weak`, `${b}Weak` ]) ] as ThemeVarPair[];
+
+  // stronger color on top of weaker color.
+  const strongOnDefault = [ ...base.map(([ a, b ]) => [ `${a}`, `${b}Strong` ]) ] as ThemeVarPair[];
+  const defaultOnWeak = [ ...base.map(([ a, b ]) => [ `${a}Weak`, `${b}` ]) ] as ThemeVarPair[];
+
+  // weak colors on top of strong colors.
+  // const weakOnDefault = [ ...base.map(([ a, b ]) => [ `${a}`, `${b}Weak` ]) ] as ThemeVarPair[];
+  // const defaultOnStrong = [ ...base.map(([ a, b ]) =>
+  //   [ `${a}Strong`, `${b}` ]) ] as ThemeVarPair[];
+
+  // CHECK: which of these pairs make sens to test?
+  return [ ...base,
+    ...strong,
+    ...weak,
+    ...strongOnDefault,
+    ...defaultOnWeak,
+    // ...weakOnDefault,
+    // ...defaultOnStrong,
+  ];
+};
+
+/** pars of theme colors used next to eachother */
+const pairs: ThemeVarPair[] = [
+  [ 'ixActive', 'ixOnActive' ],
+  [ 'ixActive', 'ixBorderActive' ],
+  [ 'ixInactive', 'ixOnInactive' ],
+  [ 'ixInactive', 'ixBorderInactive' ],
+  [ 'background', 'backgroundOn' ],
+];
+
+describe('themes', () => {
+  beforeAll(() => {
+    const newPairs = [ 'surface', 'stage', 'float' ]
+      .map(genPairs)
+      .reduce((acc, cur) => acc.concat(cur), []);
+    pairs.push(...newPairs);
+
+  });
+
+  it('should have sufficient distance between adjacent colors', () => {
+    /** defines the required minimum distance between at least one of the rgba values. */
+    const TOLERANCE = 4.1; // TODO: raise the bar.
```

Comment:
out of scope

Reply:
maybe we don't need more contrast than this. will need to verify (some) these manually